RejectedExecutionException: Trigger.getNextRunTime: null creating EJB timer

Creating a persistent EJB timer may fail with the following exception:

Caused by: java.util.concurrent.RejectedExecutionException: Trigger.getNextRunTime: null
at com.ibm.ws.concurrent.persistent.internal.PersistentExecutorImpl.newTask(PersistentExecutorImpl.java:1146)
at com.ibm.ws.concurrent.persistent.internal.PersistentExecutorImpl.schedule(PersistentExecutorImpl.java:1440)
at com.ibm.ws.ejbcontainer.timer.persistent.osgi.internal.EJBPersistentTimerRuntimeImpl.createPersistentCalendarTimer(EJBPersistentTimerRuntimeImpl.java:190)
at com.ibm.ws.ejbcontainer.osgi.internal.EJBRuntimeImpl.createPersistentCalendarTimer(EJBRuntimeImpl.java:582)
at com.ibm.ws.ejbcontainer.runtime.AbstractEJBRuntime.createTimer(AbstractEJBRuntime.java:2018)
at com.ibm.ejs.container.BeanO.createCalendarTimer(BeanO.java:2086)
at suite.timer.persistent.ejb.NoMoreTimeoutsBean.createTimer(NoMoreTimeoutsBean.java:147) ... 37 more "

--- a/dev/com.ibm.ws.ejbcontainer.timer.persistent/src/com/ibm/ws/ejbcontainer/timer/persistent/osgi/internal/PersistentTimerTaskHandlerImpl.java
+++ b/dev/com.ibm.ws.ejbcontainer.timer.persistent/src/com/ibm/ws/ejbcontainer/timer/persistent/osgi/internal/PersistentTimerTaskHandlerImpl.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2003, 2015 IBM Corporation and others.
+ * Copyright (c) 2003, 2018 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -147,6 +147,13 @@ class PersistentTimerTaskHandlerImpl extends PersistentTimerTaskHandler implemen
             nextTimeout = parsedSchedule.getNextTimeout(lastExecution.getTime());
         } else {
             nextTimeout = parsedSchedule.getFirstTimeout();
+
+            // A timer would never be scheduled if getFirstTimeout() had originally returned
+            // -1, therefore the first (and only) timeout has passed while creating the timer.
+            // In this scenario, run the timer immediately by returning the creation time.
+            if (nextTimeout == -1) {
+                return timerCreationTime;
+            }
         }
         if (nextTimeout == -1) {
             return null;

